### PR TITLE
Replace old 'important' behaviour with 'ignore_failure'

### PR DESCRIPTION
**Pull request by AdamWill**
This partly reverts commit
487fcdbd56f5e46bd569a5ffcbbbb0e90da5744a, which made openQA's
result calculation ignore the 'important' test flag entirely.

This is really inconvenient to Fedora, we found the ability to
have 'unimportant' tests quite handy and have used it in a few
different ways. Adapting to its removal has proved very hard.
So this partly restores the 'important' behaviour, but with an
'ignore_failure' flag instead of 'important'. This both inverts
the default from before, and behaves slightly differently: no
soft failure mode is involved. 'ignore_failure' does exactly
what it sounds like - a failure or soft failure of the module
will never result in failure of the overall test. If a test is
made up entirely of 'ignore_failure' modules, it will always
have 'PASSED' as its overall result. This behaviour is fine for
the cases Fedora cares about, while still allowing the removal
of the logic to handle different soft failure cases (#1293).

There will be a corresponding PR for os-autoinst, but as it
doesn't actually use the flag directly, the changes there are
quite small.

### CODE REVIEW

**Comment by AdamWill** [comment]
Note, I have no idea why the original commit added that huge bunch of modules to the fixtures. As I started this commit with a `git revert`, it removes them again, but if there's some reason for them to be there, I can put them back.

**Comment by AdamWill** [comment]
@aaannz @coolo @okurz @lnussel @garretraziel

**Comment by AdamWill** [comment]
We could even refine this to have, say, 'unimportant' and 'semiimportant', where an 'unimportant' module failing can still result in an overall pass, while a 'semiimportant' module failing would result in at best a soft fail. But not sure if it's really needed.

**Comment by coolo** [comment]
As I said in the original change, I have no problem supporting your use case. But I don't like 'unimportant' - or leave alone 'semiimportant'. 

Supporting these various states of failures in all kind of displays proved to be confusing - which was the main reason for its removal. So if you want to have test module that do not affect the result, how about having a flag that explicitly spells that instead of something quite subjective as 'unimportant'.

So have a flag ignore_result => 1

**Comment by okurz** [comment]
just to prevent the back and forth I recommend to keep the fixture modules.

I also recommend to ignore the whole notion of "important" and therefore as well the opposite as the wording matters here to confuse and cause discussions. Rather "ignore_result" as coolo proposed.

**Comment by okurz** [comment]
@Martchus

File: docs/WritingTests.asciidoc [hunk 1]
```diff
@@ -74,9 +74,11 @@ sub run {
...
 
 sub test_flags {
-    # without anything - rollback to 'lastgood' snapshot if failed
-    # 'fatal' - abort whole test suite if this fails
-    # 'milestone' - after this test succeeds, update 'lastgood'
+    # 'fatal'       - abort whole test suite if this fails (and set overall state 'failed')
+    # 'unimportant' - don't set overall state 'failed' if this fails
+    #                 if only unimportant modules fail, overall result will be 'softfailed'
+    # 'milestone'   - after this test succeeds, update 'lastgood'
+    # 'norollback'  - don't roll back to 'lastgood' snapshot if this fails
```
Comment (okurz) [inline docs/WritingTests.asciidoc]: if you introduce this as new then please do it in another commit

Comment (coolo) [inline docs/WritingTests.asciidoc]: This is just syncing with reality, this is not new.

**Comment by okurz** [comment]
even more deserves its own commit then :-)

**Comment by AdamWill** [comment]
seems like the 'unexpected coverage change' is just related to whether some kinda wait condition happens to get hit during the test run or not, nothing to do with the changes here.

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 2]
```diff
@@ -1367,8 +1374,17 @@ result in the same scenario.
 sub carry_over_bugrefs {
     my ($self) = @_;
 
-    # the carry over makes only sense for some jobs
-    return if !grep { $_ eq $self->result } (FAILED, SOFTFAILED, INCOMPLETE);
+    # the carry over makes only sense for jobs with failed modules
+    # (but including ignore_failure fails, so can't check overall result)
+    my $gotfail = 0;
+    my $modules = $self->modules;
+    while (my $m = $modules->next) {
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/Jobs.pm]: Couldn't we check failure_reason to be 'GOOD' instead of duplicating the module loop? We have to calculcate it anyway - but the calculcation is hidden in carry_over_candidate.

**Comment by coolo** [comment]
Yeah, don't worry about these 2 lines - we thought we fixed it to be deterministic, but obviously there is still some randomness.

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 2 above)
Comment (AdamWill) [inline lib/OpenQA/Schema/Result/Jobs.pm]: How about this version, then? It's OK so long as there's no other context in which we *do* want `_carray_over_candidate` to find 'GOOD' carryovers.

**Comment by garretraziel** [comment]
+1 for having just `ignore_failure` (or `ignore_result` as proposed), I think that it's sufficient for our use case and it's unambiguous.

**Comment by AdamWill** [comment]
Thanks for the merge. This is deployed now on Fedora staging and seems to be working fine in early testing.
